Interfaces for checker and IDE.

maximedenes opened the pull request:
Some files had no corresponding interfaces, so we add them and remove some dead code.

This should also fix #6177 without requiring a very dirty `Makefile` hack.

ppedrot commented:
Note that I already added an interface to `checker/check.ml` in #6276.

maximedenes commented:
Yep, I could rebase this on top of your two other `coqchk` PRs. We can merge those first if Travis ran on your branches already.

Zimmi48 commented:
The second commit message is wrong. And the fix is probably in the first commit anyways.

maximedenes commented:
> The second commit message is wrong.

Yes, thanks! Will fix.

> And the fix is probably in the first commit anyways.

No, it's when no such file remains that the bug is fixed.

SkySkimmer commented:
This leaves the question I asked in #6265 about whether we should HACKMLI in coq_makefile.

SkySkimmer commented on Makefile.ide:
```diff
@@ -106,7 +106,7 @@ $(COQIDEBYTE): $(LINKIDE)
 	        lablgtksourceview2.cma str.cma $(IDEFLAGS) $(IDECDEPSFLAGS) $^
 
-ide/coqide_main.ml: ide/coqide_main.ml4 config/Makefile # no camlp4deps here
+ide/coqide_main.ml: ide/coqide_main.mli ide/coqide_main.ml4 config/Makefile # no camlp4deps here
```
Does the generated ml really depend on the mli?

maximedenes replied:
No, that must be a mistake. Will fix.

SkySkimmer commented on checker/values.ml:
```diff
@@ -385,9 +385,6 @@ module IntMap = Map.Make(IntOrd)
 
-let register_dyn name t =
-  dyn_table := IntMap.add name t !dyn_table
-
```
What's going on here? We only expose the accessor function below, so the `dyn_table` will always be empty.

ppedrot replied:
This is a function whose use is purely speculative. It'd make possible to register data representation dynamically, e.g. if one wants to extend `votour` in interactive use. It's not used yet.

maximedenes replied:
Why not wait to have actual use cases to introduce the code?

ppedrot replied:
Yes, you can remove it. There is no way to link code into votour currently anyways.

SkySkimmer commented on ide/coqide.ml:
```diff
@@ -400,12 +400,16 @@ let reset_autosave_timer () =
 let close_and_quit = FileAux.close_and_quit
+[@@@ocaml.warning "-32"]
 let crash_save = FileAux.crash_save
+[@@@ocaml.warning "-32"]
```
What's going on here?

maximedenes replied:
Conditional compilation (used only on OSX).

SkySkimmer replied:
>Conditional compilation (used only on OSX).

But aren't they unconditionally in the mli?

maximedenes replied:
I thought no, otherwise they wouldn't trigger the warning. Now, I haven't tried to move the to the IFDEF block, maybe that's what should be done.

maximedenes replied:
No, you're right, I messed up here too. Will fix.

SkySkimmer commented on ide/coqide_main.ml4:
```diff
@@ -68,6 +69,7 @@ let set_win32_path () =
    is never read (by default) or to a temp log file (when in debug mode).
 *)
 
+[@@@ocaml.warning "-32"]
```
What's going on here?

maximedenes replied:
Conditional compilation (used only on Win32).

SkySkimmer replied:
>Conditional compilation (used only on Win32).

OK I see it.

SkySkimmer commented on ide/gtk_parsing.mli:
```diff
@@ -0,0 +1,24 @@
+                                   .. >;
+                       .. >;
+             .. > ->
+           unit
```
How about
```ocaml
val fix_double_click :
  < buffer : < get_iter : [> `INSERT ] -> GText.iter;
               move_mark : [> `INSERT | `SEL_BOUND ] ->
                 where:GText.iter -> unit;
               .. >;
    event : < connect :
                < button_press :
                    callback:([> `TWO_BUTTON_PRESS ] Gdk.event ->
                              bool) -> 'a;
                  .. >;
              .. >;
    .. > ->
  unit
```

maximedenes replied:
Right, I'll fix it.

williammpratt commented on checker/main.mli:
```diff
@@ -0,0 +1,10 @@
+(*         *       GNU Lesser General Public License Version 2.1        *)
+(************************************************************************)
+
+(* This empty file avoids a race condition that occurs when compiling a .ml file
```
Why not fix the race condition? Also, document the race condition in a repeatable way.

williammpratt commented on checker/validate.mli:
```diff
@@ -0,0 +1,9 @@
+(************************************************************************)
+(*  v      *   The Coq Proof Assistant  /  The Coq Development Team     *)
```
What's the point of all these copyright headers. You already get copyright on everything you produce in Europe anyway. You can add such headers to generated source code in a release, if you want, but adding it yourself manually just adds noise and wastes time.

Zimmi48 replied:
> What's the point of all these copyright headers.

That's the recommended practice by the FSF and it may sound a bit outdated but it is still useful to trace the origin of a single file (even if people with malign intent can very well strip this part).

> You already get copyright on everything you produce in Europe anyway.

That doesn't apply just to European countries but to all countries that signed the Berne convention.

> You can add such headers to generated source code in a release, if you want, but adding it yourself manually just adds noise and wastes time.

Right now we don't generate any source code for releases. Moreover people get the source code directly from GitHub more and more and from released tarballs less and less.

maximedenes commented:
@SkySkimmer I addressed your comments, but you may want to have a look at `coqide_main.mli`, because I hadn't realized I could not use attributes in `ml4` files to disable warnings...

SkySkimmer commented:
I don't see a coqide_main.mli, did your push fail?

maximedenes commented:
> I don't see a coqide_main.mli, did your push fail?

Sorry, something odd happened during the rebase. Now fixed.

SkySkimmer commented on ide/coqide_main.mli:
```diff
@@ -0,0 +1,17 @@
+(* These two functions are used only on Win32 platforms. Conditional compilation
+   triggers a 32 warning on other platforms, and we cannot disable the warning due
+   to the lack of attributes in camlp5. So we export them, although they are not
+   really used outside. *)
```
Instead, we could either
- put the function definitions inside the IFDEF too. This would mean they're not checked by the compiler on other platforms, but with appveyor etc that should be OK
- or put `let _ = set_win32_path (* silence warning *)` (and the other function ofc) in the ELSE branch

Not sure which of these or keeping them in the mli is better, pick what you prefer.
